libkbfs/folder_block_ops.go

```diff
@@ -3374,6 +3374,8 @@ func (fbo *folderBlockOps) updatePointer(kmd KeyMetadata, oldPtr BlockPointer, n
 				updatePointerPrefetchPriority, kmd, newPtr, block.NewEmpty(),
 				lifetime)
 		}
+		// remove any prefetches for the old pointer from the prefetcher.
+		fbo.config.BlockOps().Prefetcher().RemovePrefetch(oldPtr.ID)
```

Comment:
This seems ok, but don't we also need to call `RemovePrefetch` for blocks that have been removed?  That is the case that seems to leak blocks the worst.   So like when `folderBranchOps.notifyOneOpLocked` processes an `rmOp` or a `resolutionOp`, it should call this for each of the block IDs in `op.Unrefs()`, I think?

But I don't totally understand how the leak happened, so maybe I'm missing something.

Reply:
I think you are correct. Should it be on `rmOp` or `gcOp`, do you think?

Reply:
`rmOp` and `resolutionOp` both contain unreferenced, but archived, blocks in their `op.Unrefs()` list.  `gcOp.Unrefs()` contains blocks that are being deleted entirely.

And actually, `renameOp` can also contain `Unrefs()`, because you can rename one file over another one, deleting the latter.

So maybe the right thing to do is not be dependent on any operation type, but just do it for all `op.Unrefs()`?

Reply:
Will do. Some follow-up questions:
1. Should we then remove the `RemovePrefetch` call in `updatePointer`?
2. At what point does the root block get updated? After the `rmOp` or before?
3. Should I just be using `CancelPrefetch` in all these instances? Removing a block is a content-addressable operation...so any parent block IDs will change.

Reply:
> Should we then remove the RemovePrefetch call in updatePointer?

It still seems ok to me to keep that call.  There's no point in prefetching a block that won't be looked at anymore.  I just don't think would fix the underlying cause of the leak I was able to repro by removing a file.

> At what point does the root block get updated? After the rmOp or before?

If there was batching enabled, the root block will generally be updated while processing the `resolutionOp`, which is usually the first op in the MD update.  If for some reason there was no batching, then the root block gets updated in every operation.  (I.e., the MD update would only have one op, which could be an `rmOp`.)

> Should I just be using CancelPrefetch in all these instances? Removing a block is a content-addressable operation...so any parent block IDs will change.

I actually don't have a good grasp on the difference between remove and cancel.  If a block's removed, I guess we want any outstanding prefetches for that block canceled, and all state associated with it removed from memory?  We should make whatever calls get us that.

Reply:
OK thanks. I'll add it to all the appropriate `{resolution,rm}Op`s, and leave it in `updatePointer`.

> I actually don't have a good grasp on the difference between remove and cancel.

`RemovePrefetch` walks up the tree, subtracting the removed block's count from its parents, up to all roots of the tree. It's akin to a completion, except that it doesn't update the state of the block in the caches.

`CancelPrefetch` walks up the tree, but instead of only subtracting the removed block's count, it removes _every_ block it encounters, all the way to any roots that are connected in the current tree. I built it that way because of the content addressing I mentioned: removing any block effectively invalidates all its parents, because they would get new parents. I'm wondering if `RemoveBlock` is even a useful abstraction.

Actually now that I think about it further, once we've canceled the context for a prefetch, then the next worker that picks up that block will `CancelPrefetch` on it anyway. So we may as well just call that.

Reply:
Scratch that, it will call `CancelPrefetch`, but we will have already removed the block in order to have called the `prefetch.cancel()`, so that cancelation will be a no-op.